djangocms_versioning/models.py

```diff
@@ -1,51 +1,74 @@
-from django.core.exceptions import ImproperlyConfigured
-from django.db import connections, models
+from django.apps import apps
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
 from django.db.models import Max, Q
-from django.utils.timezone import localtime
 
+from django_fsm import FSMField, transition
 
-class BaseVersionQuerySet(models.QuerySet):
+
+class VersionQuerySet(models.QuerySet):
 
     def for_grouper(self, grouper, extra_filters=None):
         """Returns all `Version`s for specified grouper object.
 
         Additional filters can be passed via extra_filters, for example
         if version model has a FK to content object that is translatable,
-        passing `Q(content__language='en')` will return only versions
+        passing `Q(language='en')` will return only versions
         created for English language.
         """
+
         if extra_filters is None:
             extra_filters = Q()
         return self.filter(
             Q(extra_filters) &
             Q((self.model.grouper_field, grouper)),
         )
 
-    def distinct_groupers(self):
+    def distinct_groupers(self, content_model):
         """Returns a queryset of `Version` objects with unique
         grouper objects.
 
         Useful for listing, e.g. all Polls.
         """
-        if connections[self.db].features.can_distinct_on_fields:
-            return self.distinct(self.model.grouper_field).order_by('-created')
-        else:
-            inner = self.values(self.model.grouper_field).annotate(
-                Max('pk')).values('pk__max')
-            return self.filter(pk__in=inner)
+        versioning_extension = apps.get_app_config(
+            'djangocms_versioning').cms_extension
+        versionable = versioning_extension.versionables.contents[content_model]
+        ctype = ContentType.objects.get_for_model(content_model)
+        inner = content_model.objects.values(versionable.grouper_field.name).annotate(
+            Max('pk')).values('pk__max')
+        return self.filter(content_type=ctype, object_id__in=inner)
 
 
-class BaseVersion(models.Model):
+class Version(models.Model):
     # Following fields are always copied from original Version
     COPIED_FIELDS = ['label']
 
     label = models.TextField()
     created = models.DateTimeField(auto_now_add=True)
-
-    objects = BaseVersionQuerySet.as_manager()
-
-    class Meta:
-        abstract = True
+    content_type = models.ForeignKey(
+        ContentType,
+        blank=True,
+        null=True,
+        on_delete=models.SET_NULL,
+    )
+    object_id = models.PositiveIntegerField(blank=True, null=True)
+    content = GenericForeignKey('content_type', 'object_id')
+
+    objects = VersionQuerySet.as_manager()
+
+    # States
+    ARCHIVED = 'Archived'
```

Comment:
Possibly move those to `constants.py` module (and rename to VERSION_STATE_ARCHIVED/STATE_ARCHIVED/VERSION_ARCHIVED). I also prefer string choice values to be lowercase, making it easier to manually filter by specified value while debugging, what do you think of that?